### ISSUE TITLE
Rigrepdb can't process large directories

### ISSUE THREAD

**Issue by hpanago** [hpanago/rigrep#4]
Tried indexing a 1.4GB directory and it took like forever.
This should be fixed.
But first, we should identify what exactly is causing that delay with proper debugging.

@schedutron are you up for the challenge? :P

**Comment by yashacb** [comment]
Hi, 
Can I work on this?

**Comment by schedutron** [comment]
@hpanago Sure! Let's fix this!

**Comment by hpanago** [comment]
Of course, anyone can work on this.

Since we are more than 1 person looking into this, maybe we should try to organize our tasks within this thread. 

Or we can simply compare commits afterwards, I am fine with both options.

**Comment by schedutron** [comment]
@hpanago Please add appropriate labels to #5. Also, I would like to go for organising our tasks instead of comparing commits. We can advise each other instead of comparing; I think that would be a faster approach :)

**Comment by yashacb** [comment]
Organizing tasks sounds good to me too.
We need to improve the indexing time. Multi-Threading might help in this case. But then we also need to care about merging the indexes formed by different threads.
Also, I think the index can be split into multiple files, which would improve the querying time.

**Comment by schedutron** [comment]
@yashacb Excellent! Since indexing in this case is an IO-bound work, multi threading would indeed be useful. But we now need to care about how we will distribute and merge the work. I think the merge part would be especially challenging because we now need to define a functionality to merge FILES_BY_ID and REVERSE_INDEX.

**Comment by yashacb** [comment]
@schedutron Yeah, it would be challenging, but I think it is doable. 
Another thing is that the code currently reads the whole file into memory. So, I guess we can't index very large files. What if we read the file line-by-line?

**Comment by schedutron** [comment]
@yashacb Yes, I mentioned a related thing previously. Implement the line-by-line reading part for the text files to be indexed.

But how would we go about reading the pickle file also in a similar way? Whenever we lookup a token (after the indexing), the entire `.index` file is loaded into memory.

**Comment by yashacb** [comment]
I don't think we can read the pickle file in a similar way. 
But what we can do is split the whole index into parts. 
One example would be to create a index file for each alphabet. So, there will be a file called 'a.index', which contains only the words starting with 'a'. Similar is the case for other alphabets. So, for a query word, we only need to load the particular index fie, which would be (hopefully) smaller than the whole index file.
Is this clear?

**Comment by schedutron** [comment]
@yashacb Let's try it out. Indeed it will be better. Extrapolating this, what about creating a sort of tree structure, which narrows down as we go from 0th letter of the token to it's last one? This will be more efficient, right?

**Comment by yashacb** [comment]
Yeah, but I'm not sure how we would go about implementing the tree structure.

**Comment by schedutron** [comment]
@yashacb What about using directories for that?

**Comment by yashacb** [comment]
could you elaborate?

**Comment by schedutron** [comment]
@yashacb We will have a directory `a`, which will have a text file `paths.txt` containing the paths of files having token `a`. Apart from the text file, the directory will also contain subdirectories named (potentially) from `a` to `z` which will contain further text files and subdirectories. So the file for token `abc` would be  3 levels deep, it's path being `a/b/c/paths.txt`.

Thus, information for any token is in `f/o/o/paths.txt`.

**Comment by yashacb** [comment]
Sounds good, but won't the number of files become equal to the number of words(which might be pretty high)?

**Comment by schedutron** [comment]
@yashacb Number of files is not an issue if they are small enough. But since each file will have paths which other files might also have, redundancy will trouble us here. How can we go about fixing that?

**Comment by yashacb** [comment]
Can you give an example?

**Comment by hpanago** [comment]
So many awesome ideas!!

I am with you with all this, and also thought that the tree structure could help a lot, since we can exclude paths quicker. I mean that if a word goes "abg" and you happen to be searching at a branch in which first level is a, then second level is h, then you know for sure that this is not the branch you are looking for.

I will create a dev branch for us to work on all those beautiful ideas. Please commit to that for this issue.

**Comment by yashacb** [comment]
@schedutron Don't you think the tree method is equivalent to maintaining a file for each word which contains the document ids of the documents which contain the word?
So, instead of going to "a/b/c/paths.txt", we can maintain "abc.txt" instead.

**Comment by schedutron** [comment]
@yashacb But that's more linear and a bit unorganised. So you mean that if I have another word, like `abcd`, the file corresponding to it should be `abcd.txt`? And the file for every other word would be in the same folder as this file?

**Comment by yashacb** [comment]
Yeah, it is a bit unorganized.

**Comment by schedutron** [comment]
@yashacb You working on scaling this?

**Comment by schedutron** [comment]
Why have things slowed down here?

**Comment by hpanago** [comment]
Because we need to agree on a mathematical algorithm to use, and then actually code it.
